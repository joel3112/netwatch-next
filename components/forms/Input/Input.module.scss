.wrapper {
  position: relative;
  min-height: 33px;
  width: 100%;
  height: inherit;
  padding: 0 6px;
  overflow: hidden;
  background-color: transparent;
  border: 1px solid var(--color-border);
  border-radius: var(--border-radius-button);
}

.icon {
  @include absolute(top 50% left -1px);
  @include font($size: 20px, $color: var(--color-text));
  @include box(var(--size-header-action), var(--size-header-action) !important);

  border: none !important;
  background-color: transparent;
  color: var(--color-primary);
  cursor: pointer;
  transform: translateY(-50%);
  padding: 0 !important;
  margin-top: -1px;

  .disabled & {
    opacity: 0.3;
  }
}

.input {
  @include font($size: 15px, $color: var(--color-text), $weight: 400);

  font-family: var(--font-text);
  font-size: inherit;
  background-color: transparent;
  padding: 0 20px 0 4px;
  border: none;
  outline: none;
  width: 100%;
  min-height: 100%;

  .icon ~ & {
    padding-left: 25px;
  }

  .disabled & {
    color: var(--color-border) !important;
  }

  .clear ~ &[value]:not([value='']),
  .clear ~ &:focus {
    padding-right: 65px;
  }

  &::placeholder {
    opacity: 0.6;
    color: currentcolor;
  }

  &::-webkit-search-cancel-button {
    appearance: none;
    -webkit-user-modify: read-write !important;
    content: '';
  }
}

.hotKey {
  @include font($size: 13px, $color: var(--color-text), $weight: 600);

  border: 1px solid var(--color-border);
  background-color: var(--background-color-primary);
  pointer-events: none;
  border-radius: 5px;
  height: 22px;
  padding: 5px;
}
