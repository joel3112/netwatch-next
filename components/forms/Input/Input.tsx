import { useState, ChangeEvent, forwardRef, RefObject } from 'react';
import { IconType } from 'react-icons';
import { ElementHTML } from '@/types';
import { Space } from '@/components/layout';
import { Button } from '@/components/forms';
import { classes } from '@/utils/helpers';
import styles from '@/components/forms/Input/Input.module.scss';

export type InputProps = typeof defaultProps &
  ElementHTML & {
    name: string;
    id?: string;
    value?: string;
    readOnly?: boolean;
    disabled?: boolean;
    icon?: IconType;
    hotKey?: string;
    autoFocus?: boolean;
    placeholder?: string;
    autoComplete?: 'on' | 'off';
    autoCorrect?: 'on' | 'off';
    autoCapitalize?: 'on' | 'off';
    enterKeyHint?: 'go' | 'search';
    spellCheck?: 'false';
    maxLength?: number;
    type?: string;
    'aria-autocomplete'?: 'none' | 'inline' | 'list' | 'both';
    'aria-activedescendant'?: string | undefined;
    'aria-controls'?: string | undefined;
    'aria-labelledby'?: string;
    onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
    onKeyDown?: (event: KeyboardEvent) => void;
    onFocus?: (event: ChangeEvent<HTMLInputElement>) => void;
    onBlur?: () => void;
    onClick?: (event: MouseEvent) => void;
  };

const defaultProps = {
  name: ''
};

const Input = forwardRef<HTMLInputElement, InputProps>(
  (
    { className, type, name, value, placeholder, readOnly, hotKey, disabled, icon: Icon, onChange },
    ref
  ) => {
    const [inputValue, setInputValue] = useState<string>(value || '');

    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
      const { value } = event.target;

      setInputValue(value);
      onChange && onChange(event);
    };

    const handleClickIcon = () => {
      if (ref && (ref as RefObject<HTMLInputElement>).current) {
        (ref as RefObject<HTMLInputElement>).current?.focus();
      }
    };

    return (
      <Space
        align="center"
        className={classes(
          disabled && styles.disabled,
          inputValue && styles.hasValue,
          styles.wrapper,
          className
        )}>
        {Icon && (
          <Button outline ariaLabel="icon" className={styles.icon} onClick={handleClickIcon}>
            <Icon />
          </Button>
        )}

        <input
          ref={ref}
          className={classes(styles.input)}
          name={name}
          type={type}
          aria-label={name}
          placeholder={placeholder}
          autoComplete="off"
          value={inputValue}
          disabled={disabled}
          readOnly={readOnly}
          onChange={handleChange}
        />

        {hotKey && (
          <Space align="center" justify="center" className={styles.hotKey}>
            {hotKey}
          </Space>
        )}
      </Space>
    );
  }
);

Input.displayName = 'Input';
Input.defaultProps = defaultProps;

export default Input;
