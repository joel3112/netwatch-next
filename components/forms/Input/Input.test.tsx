import { fireEvent, render, screen } from '@testing-library/react';
import { BiSearch } from 'react-icons/bi';
import Input from '@/components/forms/Input/Input';

const name = 'name';
const placeholderText = 'placeholder';
const keyText = 'esc';
const handleChange = jest.fn();

describe('Tests Input component', () => {
  test('renders component correctly', () => {
    const { container } = render(<Input name={name} />);

    expect(container).toMatchSnapshot();
  });

  test('renders component correctly', () => {
    render(<Input name={name} />);

    expect(screen.getByLabelText(name)).toBeInTheDocument();
  });

  test('renders component disabled correctly', () => {
    render(<Input name={name} disabled />);

    expect(screen.getByLabelText(name)).toBeDisabled();
  });

  test('renders component with placeholder correctly', () => {
    render(<Input name={name} placeholder={placeholderText} />);

    expect(screen.getByPlaceholderText(placeholderText)).toBeInTheDocument();
  });

  test('renders component with hot key correctly', () => {
    render(<Input name={name} hotKey={keyText} />);

    expect(screen.getByText(keyText)).toBeInTheDocument();
  });

  test('renders component with value correctly', () => {
    render(<Input name={name} value="test" />);

    expect(screen.getByLabelText(name)).toHaveValue('test');
  });

  test('calls onChange prop correctly', () => {
    render(<Input name={name} onChange={handleChange} />);

    fireEvent.change(screen.getByLabelText(name), { target: { value: 'test' } });

    expect(handleChange).toHaveBeenCalled();
  });
});
