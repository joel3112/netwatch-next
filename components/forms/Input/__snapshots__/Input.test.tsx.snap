// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Input component renders component correctly 1`] = `
<div>
  <div
    class="wrapper wrapper"
  >
    <input
      aria-label="name"
      autocomplete="off"
      class="input"
      name="name"
      value=""
    />
    <style>
      
        .wrapper {
          display: flex;
          flex-direction: row;
          flex-wrap: nowrap;
          
          
          align-items: center;
        }
      
    </style>
  </div>
</div>
`;
